### ISSUE TITLE
Replace pointcloud / destroy Potree

### ISSUE THREAD

**Issue by sjelfull** [potree/potree#239]
I'm trying to achieve two things:

1. Allow the user to switch between pointclouds inside an overlay
2. Destroy Potree / the scene when it's not needed, to stop it from affecting the page performance.

I'm not able to figure out how to do either, because it doesn't look like there is methods to do so. Does Potree have the ability to do either, or do you know workarounds?

**Comment by JWGmeligMeyling** [comment]
1. You can hide the pointcloud in the material settings.
2. You can remove the pointcloud from the scene. (Multiple pointclouds should probably be rendered in the same reference frame and scene for best results anyway).

**Comment by sjelfull** [comment]
Thanks, but I'm not using the settings panel, which I assume your referring to. Is there an API method to do the same?

Is it also possible to completely destroy the viewer / Potree?

Not sure if I misunderstood your second comment, but just to clarify: I only want to show 1 pointcloud at a time. How do I do that with the API?

**Comment by JWGmeligMeyling** [comment]
The [`Potree.PointCloudOctree`](https://github.com/potree/potree/blob/master/src/PointCloudOctree.js#L36) is a [`ThreeJS.Object3D`](http://threejs.org/docs/#Reference/Core/Object3D), all of which have the property `visible` that can be set to `false` to disable rendering temporarily.

The same way the Pointcloud is added to the [`referenceFrame`](https://github.com/potree/potree/blob/master/examples/js/viewer.js#L343) you can also remove it using the `remove` method on the `referenceFrame`. This removes the Pointcloud from the `referenceFrame` and `scene`, and it won't be rendered until it is added again.

In fact, because at this point technically all references to the pointcloud should be lost, the garbage collector should pick it up, freeing the memory that was used. This however does not seem to happen completely. But at least its not eating performance.

**Comment by kingrollo** [comment]
Hi folks,

I'm trying to do something similar... I've got 5 different LiDAR scans which I'd like to display within the same container, with a separate controller for users to choose which is displaying. My first pass involved trying to destroy and recreate the entire Potree.Viewer object - should I be trying to empty it and repopulate it instead?

Anyway, with this first pass I'm finding that the second and subsequent viewers load the data in, but fail to display it on screen.

@sjelfull - did you find a working solution? Any chance of a clue?

@JWGmeligMeyling - could you possibly clarify the middle point in your last post? Are you saying we should be able to call `obj.referenceFrame.remove()` or `obj.scenePointCloud.remove()` or something else?

Thanks!
r.

**Comment by JWGmeligMeyling** [comment]
This issue is about removing a poitcloud. Showing two pointclouds simultaniously works just fine.

**Comment by kingrollo** [comment]
Yes, that's why I posted here - removing and then replacing the point cloud is exactly the method I'm trying to use. Could you possibly clarify what the correct command is?

**Comment by kingrollo** [comment]
Having looked into this further, my main problem was caused by the EDL renderer - as soon as I turned it off, things worked a lot more smoothly.

I also found it helped to destroy the DOM object, and store the requestAnimationFrame from viewer.js, and call cancelAnimationFrame(id) when dropping the point cloud. No memory leaks in Chrome.

Neither of which solved the problem, but it will do as a workaround for now.

**Comment by shrikant-dm** [comment]
I have 2D/3D toggle button from which I am switching in different views.
Now I have to completely destroy viewer object (clear all the variables) and reinitialize it.
Is there any way to achieve this functionality.

**Comment by liorGameDev** [comment]
Following... Im having the same issue....

**Comment by willemvanopstal** [comment]
@JWGmeligMeyling This is probably exactly what I need to remove a pointcloud from the scene. I tried interpreting your suggestions, but I couldn't get it to work. Is there any chance you could clarify the workflow to remove a point cloud from the scene?

_"The same way the Pointcloud is added to the referenceFrame you can also remove it using the remove method on the referenceFrame. This removes the Pointcloud from the referenceFrame and scene, and it won't be rendered until it is added again."_

**Comment by nikhilbv** [comment]
viewer.scene.pointclouds is an array that consists of the pointclouds which are present in the viewer. You can traverse the array and set the visibility to false i.e., viewer.scene.pointclouds[i].visibility=false; but this will not remove the pointclouds completely as it tells the threeJS to not to show them.
